pkg/aspect/metricsManager.go

```diff
@@ -0,0 +1,176 @@
+
+	// TODO: get descriptors from config
+	desc := []*dpb.MetricDescriptor{
+		{
+			Name:  "api_responses",
+			Kind:  dpb.COUNTER,
+			Value: dpb.INT64,
+			Labels: []*dpb.LabelDescriptor{
+				{Name: "api_method", ValueType: dpb.STRING},
+				{Name: "response_code", ValueType: dpb.INT64},
+			},
+		},
+	}
+
+	metadata := make(map[string]*metricInfo)
+	defs := make([]adapter.MetricDefinition, len(desc))
+	for i, d := range desc {
+		def := toDefinition(d)
+		metric, found := find(params.Metrics, def.Name)
+		if !found {
+			env.Logger().Warningf("No metric found for descriptor %s, skipping it", def.Name)
+			continue
+		}
+
+		defs[i] = def
+		metadata[def.Name] = &metricInfo{
+			metricKind: def.Kind,
+			value:      metric.Value,
+			labels:     metric.Labels,
+		}
+	}
+
+	asp, err := a.(adapter.MetricsBuilder).NewMetricsAspect(env, c.Builder.Params.(adapter.AspectConfig), defs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct metrics aspect with config '%v' and err: %s", c, err)
+	}
+	return &metricsWrapper{asp, metadata}, nil
+}
+
+func (*metricsManager) Kind() string                        { return MetricKind }
+func (*metricsManager) DefaultConfig() adapter.AspectConfig { return &aconfig.MetricsParams{} }
+
+func (*metricsManager) ValidateConfig(adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	// TODO: we need to be provided the metric descriptors in addition to the metrics themselves here, so we can do type assertions.
+	// We also need some way to assert the type of the result of evaluating an expression, but we don't have any attributes or an
+	// evaluator on hand.
+	return
+}
+
+func (w *metricsWrapper) Execute(attrs attribute.Bag, mapper expr.Evaluator) (*Output, error) {
+	result := &multierror.Error{}
+	var values []adapter.Value
+
+metadataLoop:
+	for name, md := range w.metadata {
+		metricValue, err := mapper.Eval(md.value, attrs)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to eval metric value for metric '%s' with err: %s", name, err))
+			continue metadataLoop // we can't satisfy this metric because we have the wrong types, so skip to the next
+		}
+
+		labels := make(map[string]interface{}, len(md.labels))
+		for label, texpr := range md.labels {
+			val, err := mapper.Eval(texpr, attrs)
+			if err != nil {
+				result = multierror.Append(result, fmt.Errorf("failed to construct value for metric '%s' with err: %s", name, err))
+				continue metadataLoop // we can't satisfy this metric because we have the wrong types, so skip to the next
+			}
+			labels[label] = val
+		}
+
+		// TODO: investigate either pooling these, or keeping a set around that has only its field's values updated.
+		// we could keep a map[metric name]value, iterate over the it updating only the fields in each value
+		values = append(values, adapter.Value{
+			Name:   name,
+			Kind:   md.metricKind,
+			Labels: labels,
+			// TODO: how do we get times?
+			StartTime:   time.Now(),
```

Comment:
at some point, there should be a timestamp attribute that is passed with data. how that gets translated, etc., is an open question.